### PR TITLE
Data prop provision for select dropdown option

### PR DESCRIPTION
**Pull request by ddouglasz**
Fixes https://github.com/commercetools/ui-kit/issues/2673

#### Summary

SelectableSearchInput: Allow choice for data-* props to be routed to search input or option dropdown. At the moment, any selector passed as a prop only appears in the search input element tree and no way to semantically capture the dropdown options using a selector. 

This PR provides a prop to pass selectors to the dropdown option.

### CODE REVIEW

**Comment by emmenko** [comment]
Thanks @ddouglasz for looking into this!

I have one main question first: from the issue request it is not clear to me if these "data attributes" are expected to be provided to each option individually or to all options.
Do you have more information about it or do we need to clarify this further?

Additionally, we might want to restrict the attributes that can be defined to only `data-` attributes, as otherwise any type of HTML attribute can be potentially provided. I don't think that's what we want here.
Btw, we have some utility function to do the filtering of `data-` attributes.

**Comment by ddouglasz** [comment]
> Thanks @ddouglasz for looking into this!
> 
> I have one main question first: from the issue request it is not clear to me if these "data attributes" are expected to be provided to each option individually or to all options. Do you have more information about it or do we need to clarify this further?
> 
> Additionally, we might want to restrict the attributes that can be defined to only `data-` attributes, as otherwise any type of HTML attribute can be potentially provided. I don't think that's what we want here. Btw, we have some utility function to do the filtering of `data-` attributes.

Very important observations, I will get more clarity on these. 
Thank you 🙏🏾

**Comment by ddouglasz** [comment]
> I have one main question first: from the issue request it is not clear to me if these "data attributes" are expected to be provided to each option individually or to all options. Do you have more information about it or do we need to clarify this further?

So I was able to get a [response](https://github.com/commercetools/ui-kit/issues/2673#issuecomment-1880712440) the requirement is to have 1 data attribute attached to the dropdown component itself.

**Comment by ddouglasz** [comment]
@emmenko  Updated all feedback here: https://github.com/commercetools/ui-kit/pull/2678/commits/158e3f212ab003973a195a88f17a83bad8bd30fd

Thank you for the pointers

File: packages/components/inputs/selectable-search-input/src/selectable-search-input.tsx [hunk 1]
```diff
@@ -206,6 +206,10 @@ export type TSelectableSearchInputProps = {
    * Show submit button in the input
    */
   showSubmitButton?: boolean;
+  /**
+   * array of selector options that can be used to pass `data-*` props selectors to the options dropdown
+   */
+  optionSelector?: TOptionSelector[];
```
Comment (CarlosCortizasCT) [inline packages/components/inputs/selectable-search-input/src/selectable-search-input.tsx:212]: I think the name of this prop might be misleading as you should be able to pass any `data-*` attribute and also would be difficult, from the consumer point of view, to understand how those attributes can be forwarded to the select or to the input.

My proposal would be to create two new props for this purpose:
* inputDataProps
* selectDataProps

We can also keep the raw `data-*` props we're currently receiving and forwarding to the `input` element for backwards compatibility.

This way, consumers who want to provide `data` attributes would need to explicitly indicate where they want them to be forwarded.

Example:
```ts
<SelectableSearchInput
  ...
  selectDataProps={{
    'tracking': 'products-search-type'
  }}
  inputDataProps={{
    'tracking': 'product-search-value' 
  }}
  ...
>
```

Then we would transform those values to HTML `data` attributes in the corresponding elements.

@emmenko would that work for you?

Comment (ddouglasz) [reply packages/components/inputs/selectable-search-input/src/selectable-search-input.tsx:212]: This makes so much sense. I will refactor once we conclude on this approach

Thank you 🙏🏾

Comment (emmenko) [reply packages/components/inputs/selectable-search-input/src/selectable-search-input.tsx:212]: Makes sense to me, thanks for the suggestion! 🤗

Comment (ddouglasz) [reply packages/components/inputs/selectable-search-input/src/selectable-search-input.tsx]: updated here: https://github.com/commercetools/ui-kit/pull/2678/commits/79399cc59a7f211ce9b969875c4b515f8b6c24fe

Thank you

File: packages/components/inputs/selectable-search-input/docs/usage-example.js [hunk 2]
```diff
@@ -15,6 +15,14 @@ const Example = () => (
       { value: 'one', label: 'One' },
       { value: 'two', label: 'Two' },
     ]}
+    selectDataProps={[
+      { 'data-select-1': 'value-1' },
+      { 'data-select-2': 'value-2' },
+    ]}
+    inputDataProps={[
+      { 'data-input-1': 'value-1' },
```
[the remark this answers is not in the document]
Comment (ddouglasz) [reply packages/components/inputs/selectable-search-input/docs/usage-example.js:23]: Thank you for the feedback. So if I get this clearly, if a user passes in `inputDataProps={[{ 'prop-1': 'value-1' },]}` then we should internally convert it to something like `<input data-prop-1="value-1" />` ? so user does not necessarily need to add the `data-` prefix ?

[the remark this answers is not in the document]
Comment (CarlosCortizasCT) [reply packages/components/inputs/selectable-search-input/docs/usage-example.js:23]: Yes, that's what I would expect.

[the remark this answers is not in the document]
Comment (ddouglasz) [reply packages/components/inputs/selectable-search-input/docs/usage-example.js]: updated here: https://github.com/commercetools/ui-kit/pull/2678/commits/7bdd6af91ee5c5052752dc28fb049b10585e2818

Thank you

File: packages/components/inputs/selectable-search-input/src/selectable-search-input.tsx [hunk 3]
```diff
@@ -239,6 +247,28 @@ const SelectableSearchInput = (props: TSelectableSearchInputProps) => {
   const containerRef = useRef<HTMLDivElement>(null);
   const textInputRef = useRef<HTMLInputElement>(null);
 
+  const getDataProps = (data?: TDataProps[]) => {
+    const addPrefixToKeys = (obj: Record<string, string>) => {
```
Comment (ddouglasz) [inline packages/components/inputs/selectable-search-input/src/selectable-search-input.tsx:251]: Thinking too make this a util since its used in 2 places 🤔

**Comment by CarlosCortizasCT** [comment]
Thanks for all the work in this PR, @ddouglasz 🙇 

I think that maybe we weren't aligned on the path to take in terms of what I was suggesting so I created [this alternative PR](https://github.com/commercetools/ui-kit/pull/2684) where I tried to show an alternative that might help use to get to a common ground.

Please let me know your thoughts 🙏 

We can have a sync if you prefer to 👍

**Comment by ddouglasz** [comment]
> Thanks for all the work in this PR, @ddouglasz 🙇
> 
> I think that maybe we weren't aligned on the path to take in terms of what I was suggesting so I created [this alternative PR](https://github.com/commercetools/ui-kit/pull/2684) where I tried to show an alternative that might help use to get to a common ground.
> 
> Please let me know your thoughts 🙏
> 
> We can have a sync if you prefer to 👍

I have gone through the work now, I see what you mean. It makes sense to me. 
Thank you very much 🙏🏾

**Review by chloe0592 — APPROVED** [review]
Makes sense to me. Thanks for the work and effort @ddouglasz 🙏

File: .changeset/spicy-books-invite.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@commercetools-uikit/selectable-search-input': patch
+---
+
+provide option to pass selector prop to the dropdown option in selectable search input.
```
[the remark this answers is not in the document]
Comment (ddouglasz) [reply .changeset/spicy-books-invite.md:5]: I will raise a small follow up PR to clean these up. 
Thank you

**Pull request merged**